runtime/lua/vim/diagnostic.lua

```diff
@@ -715,6 +715,12 @@ function M.set(namespace, bufnr, diagnostics, opts)
   end
 
   if vim.api.nvim_buf_is_loaded(bufnr) then
+    vim.api.nvim_buf_attach(bufnr, false, {
+      on_detach = function()
+        rawset(diagnostic_cache, bufnr, nil) -- clear cache
+      end,
+    })
+
```

Comment:
This will just call `nvim_buf_attach` twice for loaded buffers then, as it is also called in the `__index` function for the diagnostic cache's `metatable`.

What if instead we just force buffers to load when they are added to the diagnostic cache? I wonder what the performance cost of that would be?

Reply:
@gpanders `use vim.fb.buload` doesn't fix that. it will load once of current buffer.like i said it not trigger on `__index`

Reply:
>@gpanders use vim.fb.buload doesn't fix that.

Did you try it?

Reply:
yup. still has error

Reply:
done. I forget restore the `$VIMRUNTIME` .so it not take effect. confirm work. I forget this api can load buffer. my mistake

Reply:
Please add the comment from the patch above:

```lua
-- Force buffer to load, nvim_buf_attach does not work on unloaded buffers
```

Reply:
> What if instead we just force buffers to load when they are added to the diagnostic cache? I wonder what the performance cost of that would be?

This would trigger all autocmds right? If so I think this could be potentially very costly. Many language servers report diagnostcs for the full projects which can easily result in 10k+ diagnostics spanning several hundred files.

Reply:
That's a good point. I suppose we could disable autocommands before loading the buffer, but that probably has *other* side effects.

Let's try using a `BufDelete` autocmd instead of an `on_detach` callback in `nvim_buf_attach`. That should still fire even for buffers that were never loaded.

Reply:
I had think about it。but dont find a good place to write it

Reply:
done.